File: app/src/main/java/tomaszkopacz/meetbam/presenters/AccountPhotosFragmentPresenter.java
package tomaszkopacz.meetbam.presenters;

import android.support.v7.widget.RecyclerView;
import android.view.View;

import com.bumptech.glide.Glide;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import tomaszkopacz.meetbam.activities.AccountPhotosFragment;
import tomaszkopacz.meetbam.model.Post;
import tomaszkopacz.meetbam.service.LoginService;
import tomaszkopacz.meetbam.service.PostAdapter;
import tomaszkopacz.meetbam.service.PostTimeProvider;
import tomaszkopacz.meetbam.service.PostViewHolder;
import tomaszkopacz.meetbam.service.WebService;

/**
 * Created by tomas on 02.03.2018.
 */

public class AccountPhotosFragmentPresenter implements RecyclerViewPresenter{

    private AccountPhotosFragment fragment;
    private WebService mWebService;
    private LoginService mLoginService;

    // photos list
    private PostAdapter adapter;
    private List<Post> photos = new ArrayList<>();

    // photo base url
    private static final String BASE_URL = "http://meetbam.cba.pl/";

    /**
     * Constructor.
     */
    public AccountPhotosFragmentPresenter(AccountPhotosFragment fragment, WebService service){
        this.fragment = fragment;
        this.mWebService = service;
        this.mLoginService = new LoginService(fragment.getActivity().getApplicationContext());
        this.adapter = new PostAdapter(this);
    }

    /**
     * Downloads posts and send them to activity.
     */
    public void downloadPostsList(){

        // remove previous elements
        photos.clear();

        // when the elements downloaded
        Call call = mWebService.getUserPosts(mLoginService.getUserMail());
        call.enqueue(new Callback<List<Post>>() {

            @Override
            public void onResponse(Call<List<Post>> call, Response<List<Post>> response) {

                // assign response objects to list
                photos = response.body();

                // send posts to activity
                fragment.putPhotos(adapter);
            }

            @Override
            public void onFailure(Call<List<Post>> call, Throwable t) {

            }
        });
    }

    @Override
    public void onItemBoundAtPosition(RecyclerView.ViewHolder holder, int position) {
        Post post = photos.get(position);

        //set text views
        ((PostViewHolder) holder).getName1().setText(post.getName1() + " " + post.getSurname1());
        ((PostViewHolder) holder).getName2().setText(post.getName2() + " " + post.getSurname2());

        // download and set image
        Glide
                .with(fragment)
                .load(BASE_URL + post.getPhotoDir())
                .into(((PostViewHolder) holder).getPhoto());

        // count time since post was uploaded
        float timeAgo = PostTimeProvider.countTimeAgo(post.getTime());

        // get appropriate info about how long ago post was uploaded
        String timeAgoText = PostTimeProvider.getTimeAgoText(timeAgo);
        ((PostViewHolder) holder).getTime().setText(timeAgoText);
    }

    @Override
    public void onItemClick(View view) {

    }

    @Override
    public int getItemCount() {
        return photos.size();
    }
}

File: app/src/main/java/tomaszkopacz/meetbam/presenter/AccountPhotosFragmentPresenter.kt
package tomaszkopacz.meetbam.presenter

import android.support.v7.widget.RecyclerView
import android.view.View
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import tomaszkopacz.meetbam.entity.Post
import tomaszkopacz.meetbam.interactor.WebService
import tomaszkopacz.meetbam.view.AccountPhotosFragment
import tomaszkopacz.meetbam.view.MainApp
import tomaszkopacz.meetbam.view.PostAdapter
import java.util.*
import javax.inject.Inject

class AccountPhotosFragmentPresenter(private val fragment: AccountPhotosFragment)
    : RecyclerViewPresenter {

    //service
    @Inject lateinit var webService: WebService
    val app = fragment.activity!!.application as MainApp

    //posts
    private var adapter = PostAdapter(this)
    private var photos: MutableList<Post>? = ArrayList()

    companion object {
        private val BASE_URL = "http://meetbam.cba.pl/"
    }

    init {
        (fragment.activity!!.application as MainApp)
                .component!!.inject(this)
    }

    fun downloadPostsList() {

        photos!!.clear()

        val call = webService.getUserPosts(app.getLoggedUser().mail!!)
        call.enqueue(object : Callback<List<Post>> {

            override fun onResponse(call: Call<List<Post>>, response: Response<List<Post>>) {

                photos = response.body() as MutableList<Post>
                fragment.putPhotos(adapter!!)
            }

            override fun onFailure(call: Call<List<Post>>, t: Throwable) {

            }
        })
    }

    override fun onItemBoundAtPosition(holder: RecyclerView.ViewHolder, position: Int) {
        val post = photos!![position]
        holder as PostAdapter.PostViewHolder
        adapter.setContent(fragment.context!!, holder, post, BASE_URL)
    }

    override fun onItemClick(view: View) {

    }

    override fun onLongItemClick(view: View) {

    }

    override fun itemCount(): Int {
        return photos!!.size
    }
}
